### PR TITLE
Fix keycode mappings for via and ensure they don't change within protocol

### PR DESCRIPTION
**Pull request by xyzz**
## Description

Add checks to ensure via keycodes stay the same.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #11157

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/via_ensure_keycode.h [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+_Static_assert(KC_WH_L == 251, "");
+_Static_assert(KC_WH_R == 252, "");
+_Static_assert(KC_ACL0 == 253, "");
+_Static_assert(KC_ACL1 == 254, "");
+_Static_assert(KC_ACL2 == 255, "");
+_Static_assert(KC_LCAP == 130, "");
+_Static_assert(KC_LNUM == 131, "");
+_Static_assert(KC_LSCR == 132, "");
+
+_Static_assert(MACRO00 == 0x5F12, "");
```
Comment (drashna) [inline quantum/via_ensure_keycode.h:256]: what about USER00-USER15?  As well as FN_MO13, and FN_MO23? 

```suggestion
_Static_assert(FN_MO13 == 0x5F10, "");
_Static_assert(FN_MO23 == 0x5F11, "");
_Static_assert(MACRO00 == 0x5F12, "");
```

File: quantum/via_ensure_keycode.h [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+_Static_assert(MACRO06 == 0x5F18, "");
+_Static_assert(MACRO07 == 0x5F19, "");
+_Static_assert(MACRO08 == 0x5F1A, "");
+_Static_assert(MACRO09 == 0x5F1B, "");
+_Static_assert(MACRO10 == 0x5F1C, "");
+_Static_assert(MACRO11 == 0x5F1D, "");
+_Static_assert(MACRO12 == 0x5F1E, "");
+_Static_assert(MACRO13 == 0x5F1F, "");
+_Static_assert(MACRO14 == 0x5F20, "");
+_Static_assert(MACRO15 == 0x5F21, "");
```
Comment (drashna) [inline quantum/via_ensure_keycode.h:271]:
```suggestion
_Static_assert(MACRO15 == 0x5F21, "");

_Static_assert(USER00 == 0x5F80, "");
_Static_assert(USER01 == 0x5F81, "");
_Static_assert(USER02 == 0x5F82, "");
_Static_assert(USER03 == 0x5F83, "");
_Static_assert(USER04 == 0x5F84, "");
_Static_assert(USER05 == 0x5F85, "");
_Static_assert(USER06 == 0x5F86, "");
_Static_assert(USER07 == 0x5F87, "");
_Static_assert(USER08 == 0x5F88, "");
_Static_assert(USER09 == 0x5F89, "");
_Static_assert(USER10 == 0x5F8A, "");
_Static_assert(USER11 == 0x5F8B, "");
_Static_assert(USER12 == 0x5F8C, "");
_Static_assert(USER13 == 0x5F8D, "");
_Static_assert(USER14 == 0x5F8E, "");
_Static_assert(USER15 == 0x5F8F, "");
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: quantum/via_ensure_keycode.h (hunk 1 above)
Comment (xyzz) [reply quantum/via_ensure_keycode.h:256]: these are up top but i'll move them down

**Review by drashna — APPROVED** [review]

**Comment by xyzz** [comment]
In draft until we can get "official" via keycodes list from @Wilba6582

**Comment by infinityis** [comment]
I compared the latest list of keycodes as of commit 67c18d0 to the list of keycodes I found in via source. I added all the ones that were missing, they start on line 292 here:
https://github.com/customMK/qmk_firmware/blob/f69a2214d3b7f589a58ddc712739489c3480dd78/quantum/via_ensure_keycode.h

I tried to set this up as a pull request to xyzz's fork of qmk_firmware, but it doesn't show up the list of base repositories to choose from when doing a pull request. I also don't think there is a way I can commit directly to this pull request either, so...copy and paste?

**Comment by xyzz** [comment]
I'm not sure where these came from originally, as I tried compiling your commit and these keycodes aren't in qmk:

```
quantum/via_ensure_keycode.h:349:16: error: 'FN_TT13' undeclared here (not in a function); did you mean 'FN_MO13'?
  349 | _Static_assert(FN_TT13 == 12081, "");
      |                ^~~~~~~
      |                FN_MO13
quantum/via_ensure_keycode.h:349:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:350:16: error: 'FN_TT23' undeclared here (not in a function); did you mean 'FN_MO23'?
  350 | _Static_assert(FN_TT23 == 12082, "");
      |                ^~~~~~~
      |                FN_MO23
quantum/via_ensure_keycode.h:350:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:351:16: error: 'SPC_FN1' undeclared here (not in a function); did you mean 'KC_FN1'?
  351 | _Static_assert(SPC_FN1 == 16684, "");
      |                ^~~~~~~
      |                KC_FN1
quantum/via_ensure_keycode.h:351:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:352:16: error: 'SPC_FN2' undeclared here (not in a function); did you mean 'KC_FN2'?
  352 | _Static_assert(SPC_FN2 == 16940, "");
      |                ^~~~~~~
      |                KC_FN2
quantum/via_ensure_keycode.h:352:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:353:16: error: 'SPC_FN3' undeclared here (not in a function); did you mean 'KC_FN3'?
  353 | _Static_assert(SPC_FN3 == 17196, "");
      |                ^~~~~~~
      |                KC_FN3
quantum/via_ensure_keycode.h:353:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:370:16: error: 'BR_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  370 | _Static_assert(BR_INC == 24320, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:370:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:371:16: error: 'BR_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  371 | _Static_assert(BR_DEC == 24321, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:371:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:372:16: error: 'EF_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  372 | _Static_assert(EF_INC == 24322, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:372:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:373:16: error: 'EF_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  373 | _Static_assert(EF_DEC == 24323, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:373:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:374:16: error: 'ES_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  374 | _Static_assert(ES_INC == 24324, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:374:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:375:16: error: 'ES_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  375 | _Static_assert(ES_DEC == 24325, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:375:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:376:16: error: 'H1_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  376 | _Static_assert(H1_INC == 24326, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:376:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:377:16: error: 'H1_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  377 | _Static_assert(H1_DEC == 24327, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:377:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:378:16: error: 'S1_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  378 | _Static_assert(S1_INC == 24328, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:378:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:379:16: error: 'S1_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  379 | _Static_assert(S1_DEC == 24329, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:379:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:380:16: error: 'H2_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  380 | _Static_assert(H2_INC == 24330, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:380:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:381:16: error: 'H2_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  381 | _Static_assert(H2_DEC == 24331, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:381:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:382:16: error: 'S2_INC' undeclared here (not in a function); did you mean 'BL_INC'?
  382 | _Static_assert(S2_INC == 24332, "");
      |                ^~~~~~
      |                BL_INC
quantum/via_ensure_keycode.h:382:16: error: expression in static assertion is not an integer
quantum/via_ensure_keycode.h:383:16: error: 'S2_DEC' undeclared here (not in a function); did you mean 'BL_DEC'?
  383 | _Static_assert(S2_DEC == 24333, "");
      |                ^~~~~~
      |                BL_DEC
quantum/via_ensure_keycode.h:383:16: error: expression in static assertion is not an integer
```

**Comment by infinityis** [comment]
I also don't know where they come from originally, but I do know that some of the labeling is different between your list and what I found in via, which may explain the discrepancy. You can see the full comparison here, where (for example) your list has "KC_ENTER" and via has "KC_ENT":
https://docs.google.com/spreadsheets/d/1owrKR5EnsxGsORwSW9IY6EYWPr1mYfAlo7_u4HbjAxg/edit?usp=sharing

If you're asking how I got the key codes from via source, it's just that via is an electron app, so under the hood, via is just executing uncompiled javascript sitting on top of a dedicated browser-like rendering engine. The uncompiled javascript is just as accessible as the javascript running on any web page you visit (noting however that in a similar vein, you still won't have legal rights to copy or distribute the code without it being properly licensed). It's fairly straightforward to extract the javascript from most any electron app (see here https://medium.com/how-to-electron/how-to-get-source-code-of-any-electron-application-cbb5c7726c37). Once you've done that, you can find the matching list of key codes in /app/dist/renderer.prod.js

It's pretty neat to poke around in the via code and see interesting things like how there really is dedicated code in there for Satisfaction 75 to handle dynamic encoder assignments and dedicated menus. Looking through it also yields some insight into why it is likely being run as a closed source project. For example, the encoder code for Satisfaction 75 surely works, but it is written exclusively for that keyboard, and not in a scalable, generalized way that other keyboards could take advantage of (yet, anyway). Implementing such features in a public, open source repository is usually going to be inherently slower, involving consensus, abstractions, resolving public issues, etc. instead of being focused on providing a responsive service to smaller number of paying customers (in this case, keyboard manufacturers)

**Comment by xyzz** [comment]
Okay I went ahead and removed invalid keycodes from the list, so now it's ready for review

**Review by tzarc — APPROVED** [review]

**Pull request merged**